src/backend/replication/walproposer.c

```diff
@@ -383,17 +379,27 @@ HandleWalKeeperResponse(void)
 		 * read from WAL and send are plain sheets of bytes, but safekeepers
 		 * ack only on commit boundaries.
 		 */
-		if (msg->req.endLsn >= minQuorumLsn && minQuorumLsn != InvalidXLogRecPtr)
+		if (msgQueueAck->req.endLsn >= minQuorumLsn && minQuorumLsn != InvalidXLogRecPtr)
 		{
 			truncateLsn = minQuorumLsn;
 			candidateTruncateLsn = InvalidXLogRecPtr;
 		}
-		else if (msg->req.endLsn >= candidateTruncateLsn &&
+		else if (msgQueueAck->req.endLsn >= candidateTruncateLsn &&
 				 candidateTruncateLsn != InvalidXLogRecPtr)
 		{
 			truncateLsn = candidateTruncateLsn;
 			candidateTruncateLsn = InvalidXLogRecPtr;
 		}
+
+		msgQueueAck = msgQueueAck->next;
+	}
+
+	/* Cleanup message queue up to truncateLsn, but only messages received by everyone */
+	while (msgQueueHead != NULL && msgQueueHead->ackMask == ((1 << n_walkeepers) - 1) && msgQueueHead->req.endLsn <= truncateLsn)
```

Comment:
Isn't condition `msgQueueHead->ackMask == ((1 << n_walkeepers) - 1)` here unneccessary? Any chunk beyond truncateLsn must be already received by everyone.

Reply:
It is necessary. I also thought it's not needed, but it seems that code without it fails with 'Use After Free' error in the following case. A bunch of zero-length messages with `beginLsn == endLsn` are being broadcasted to the safekeepers, but not all of them are acknowledged back. After that `truncateLsn` is moved to this exact lsn and without this condition all empty messages are freed, but they're still not acked and used to retry sending.

Reply:
Oh, ok.